packages/core/src/parser/util.ts

```diff
@@ -506,6 +506,71 @@ export function stopBefore<N extends Returnable>(
 	}
 }
 
+/**
+ * @param terminators A list of characters the parser will stop at if it finds them
+ * before a backslash
+ * @returns A parser that is based on the passed-in `parser`, but concatenates lines
+ * together when we reach, in order:
+ * - a backslash
+ * - whitespace (optional)
+ * - a newline
+ * - whitespace (optional)
+ */
+export function concatOnTrailingBackslash<N extends Returnable>(
+	parser: InfallibleParser<N>,
+	terminators: string[],
+): InfallibleParser<N>
+export function concatOnTrailingBackslash<N extends Returnable>(
+	parser: Parser<N>,
+	terminators: string[],
+): Parser<N>
+export function concatOnTrailingBackslash<N extends Returnable>(
+	parser: Parser<N>,
+	terminators: string[],
+): Parser<N> {
+	return (src, ctx): Result<N> => {
+		let wrappedStr = src.sliceToCursor(0)
+		const wrapper = '\\'
+		const wrappedSrcCursor = wrappedStr.length
+		const indexMap: IndexMap = []
+
+		while (src.canRead() && !terminators.includes(src.peek())) {
+			wrappedStr += src.readUntil(wrapper, ...terminators)
+			if (!src.canRead() || terminators.includes(src.peek())) {
+				// Stop wrapping `src` if we reach end of file or a terminator before a wrapper
+				break
+			}
+
+			// If we get here, then `src.cursor` is at a wrapper
+			if (src.hasNonSpaceAheadInLine(1)) {
+				wrappedStr += src.read()
+				continue
+			}
+
+			// Create an index map that skips the trailing backslash + whitespace
+			const from = src.getCharRange()
+			// skip the `\`
+			src.skip()
+			// skip trailing whitespace + newline + preceding whitespace on the next line
+			src.skipWhitespace()
+			const to = src.getCharRange(-1)
+			indexMap.push({
+				inner: Range.create(wrappedStr.length),
+				outer: Range.span(from, to),
+			})
+		}
+
+		const wrappedSrc = new Source(
+			wrappedStr,
+			IndexMap.merge(src.indexMap, indexMap),
+		)
```

Comment:
I suppose Index Map is sth that when you report a range it maps it accordingly? That's neat!

Reply:
yeah this took me a bit to understand. basically it's just saying "here's what the parser will see, but here's what character ranges that maps to in vscode's actual text document"